### PR TITLE
MAINT: Add BarReader base class for both minute and session readers

### PR DESCRIPTION
**Pull request by jbredeche**
No functional changes.

### CODE REVIEW

File: zipline/data/bar_reader.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
```
Comment (ehebert) [inline zipline/data/bar_reader.py]: Missing copyright notice.

File: zipline/data/bar_reader.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    @abstractmethod
+    def get_value(self, sid, dt, field):
+        """
+        Retrieve the value at the given coordinates.
+
+        Parameters
+        ----------
+        sid : int
+            The asset identifier.
+        session : pd.Timestamp
```
Comment (ehebert) [inline zipline/data/bar_reader.py]: Now that it is generalized, this should be `dt` not `session`.

File: zipline/data/bar_reader.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        ----------
+        asset : zipline.asset.Asset
+            The asset for which to get the last traded minute.
+        dt : pd.Timestamp
+            The minute at which to start searching for the last traded minute.
+
+        Returns
+        -------
+        last_traded : pd.Timestamp
+            The minute of the last trade for the given asset, using the input
```
Comment (ehebert) [inline zipline/data/bar_reader.py]: `s/minute/dt/g`

File: zipline/data/resample.py [hunk 4]
```diff
@@ -525,6 +525,9 @@ def last_available_dt(self):
     def first_trading_day(self):
         return self._minute_bar_reader.first_trading_day
 
+    def get_last_traded_dt(self, asset, dt):
```
Comment (ehebert) [inline zipline/data/resample.py]: Could you add a basic test to `test_resample` to cover this?

e.g.:

```
    def test_get_last_traded_dt(self):
        self.assertEqual(self.END_DATE,
                                   self.session_bar_reader.get_last_traded_dt)
```

Comment (jbredeche) [inline zipline/data/resample.py]: added

File: zipline/data/bar_reader.py (hunk 1 above)
Comment (jbredeche) [inline zipline/data/bar_reader.py]: no longer!

**Pull request merged**